airbyte-metrics/lib/src/main/java/io/airbyte/metrics/lib/DogStatsDMetricSingleton.java

```diff
@@ -0,0 +1,159 @@
+    if (instance == null) {
+      throw new RuntimeException("You must initialize configuration with the initialize() method before getting an instance.");
+    }
+    return instance;
+  }
+
+  public synchronized static void initialize(final AirbyteApplication app, final boolean publish) {
+    initialize(app, publish, null, null);
+  }
+
+  @VisibleForTesting
+  public synchronized static void initialize(final AirbyteApplication app, final boolean publish, final String ddAgentHost, final String ddPort) {
+    if (instance != null) {
+      throw new RuntimeException("You cannot initialize configuration more than once.");
+    }
+
+    if (!publish) {
+      log.info("Starting stub DogstatsD client..");
+      instance = new DogStatsDMetricSingleton();
+      return;
+    }
+
+    log.info("Starting DogStatsD client..");
+    if (!ddAgentHost.isBlank() || !ddPort.isBlank()) {
+      instance = new DogStatsDMetricSingleton(app.getApplicationName(), publish, ddAgentHost, ddPort);
+      return;
+    }
+
+    instance = new DogStatsDMetricSingleton(app.getApplicationName(), publish);
+  }
+
+  @VisibleForTesting
+  public synchronized static void flush() {
+    instance = null;
+  }
+
+  private DogStatsDMetricSingleton(final String appName, final boolean publish) {
+    instancePublish = publish;
+    statsDClient = new NonBlockingStatsDClientBuilder()
+        .prefix(appName)
+        .hostname(System.getenv("DD_AGENT_HOST")) // Matches Airbyte Cloud Datadog env vars.
+        .port(Integer.parseInt(System.getenv("DD_DOGSTATSD_PORT"))) // Matches Airbyte Cloud Datadog env vars.
+        .build();
+  }
+
+  private DogStatsDMetricSingleton(final String appName, final boolean publish, final String ddAgentHost, final String ddPort) {
+    instancePublish = publish;
+    statsDClient = new NonBlockingStatsDClientBuilder()
+        .prefix(appName)
+        .hostname(ddAgentHost)
+        .port(Integer.parseInt(ddPort))
+        .build();
+  }
+
+  private DogStatsDMetricSingleton() {
+    instancePublish = false;
+    statsDClient = null;
+  }
+
+  /**
+   * Increment or decrement a counter.
+   *
+   * @param metric
+   * @param amt to adjust.
+   * @param tags
+   */
+  public void count(final AirbyteMetricsRegistry metric, final double amt, final String... tags) {
+    if (instancePublish) {
+      log.info("publishing count, name: {}, value: {}", metric.metricName, amt);
+      statsDClient.count(metric.metricName, amt, tags);
+    }
+  }
+
+  /**
+   * Record the latest value for a gauge.
+   *
+   * @param metric
+   * @param val to record.
+   * @param tags
+   */
+  public void gauge(final AirbyteMetricsRegistry metric, final double val, final String... tags) {
+    if (instancePublish) {
+      log.info("publishing gauge, name: {}, value: {}", metric, val);
+      statsDClient.gauge(metric.metricName, val, tags);
+    }
+  }
+
+  /**
+   * Submit a single execution time aggregated locally by the Agent. Use this if approximate stats are
```

Comment:
as a developer reading this, it's not clear what we mean for approximate versus precise stats and when i should use which method. can you add more clarity, please?

Reply:
I've updated this. Is this clearer?